### ISSUE TITLE
How to Diff non-differentiable Kornia Ops?

### ISSUE THREAD

**Issue by xp1632** [xp1632/Aalen_working_log#75]
# Differentiable Operations — Design Notes

### Goal 
- We want to achieve segmentation tasks and other more complicated computer vision tasks with Kornia operators and with our gradient descent optimizer.
- To do this, we need **all operations in the pipeline to be differentiable** — gradients must flow through every parameter so the optimizer can tune them via backpropagation.

**Comment by xp1632** [comment]
### Constraints

#### 1. Kornia operator side: Not all operators are gradient-compatible

Not all Kornia operators use continuous, differentiable parameters. Some detach key parameters from the computation graph, and some have no learnable parameters at all.

| Original Kornia Op | Learnable Params | Gradient Issue |
|---|---|---|
| `GaussianBlur2D` | `radius` [1.0, 10.0] | sigma is `.detach()`ed internally — **gradients don't flow through blur strength** |
| `UnsharpMask` | `radius` [1.0, 10.0] | Same `.detach()` issue as GaussianBlur2D |
| `MorphologyOperator` | **None** — SE is `torch.ones(5,5)` in `non_diff_params` | Binary structuring element, not a `nn.Parameter` at all. Even if made a parameter, Kornia's morphology doesn't propagate gradients through the SE kernel |
| `RgbToGrayscale` | **None** — fixed weights 0.299/0.587/0.114 | No learnable parameters, channel mixing is hardcoded |
| Thresholding | **N/A** — no Kornia differentiable threshold op | Hard step function kills gradients entirely |
| `AdjustBrightness` | `factor` | Kornia's implementation breaks gradient chain (known issue) |
| `AdjustContrast` | `factor` | Same gradient chain issue |
| `AdjustGamma` | `gamma`, `gain` | Same gradient chain issue |
| `AdjustSaturation` | `factor` | Same gradient chain issue |

#### 2. Chaldene side: Animation and parameter mismatch

During the optimization animation, we replay snapshots of intermediate parameters on Chaldene visual nodes. The problem: **Diff ops use different parameter names/formats than the Kornia NodeSpecs in Chaldene**.

| Diff Op (Optimizer) | Diff Params | Kornia Chaldene Node | Kornia Params | Mismatch |
|---|---|---|---|---|
| `DiffGaussianBlur` | `sigma` [0.1, 10] | `gaussian_blur2d` | `kernel_height`, `kernel_width`, `sigma_y`, `sigma_x` | Need to derive kernel_size from sigma |
| `DiffUnsharpMask` | `sigma` [0.1, 10] | `unsharp_mask` | `kernel_height`, `kernel_width`, `sigma_y`, `sigma_x` | Same conversion needed |
| `DiffSobel` | (none) | `sobel` | (none) | ✅ No issue |
| `DiffMorphology` | `se_values` (25 floats) | `dilation`/`erosion`/etc. | `kernel` (binary image) | SE values can't map to a simple slider |
| `DiffThreshold` | `threshold`, `steepness` | No Kornia equivalent | — | No mapping possible |
| `DiffRgbToGrayscale` | `weights` (3 floats) | `rgb_to_grayscale` | (none) | Kornia version has no params |
| `DiffBrightness` | `factor` | `diff_brightness` | `factor` | ✅ Direct (already solved) |
| `DiffContrast` | `factor` | `diff_contrast` | `factor` | ✅ Direct (already solved) |
| `DiffGamma` | `gamma` | `diff_gamma` | `gamma` | ✅ Direct (already solved) |
| `DiffSaturation` | `factor` | `diff_saturation` | `factor` | ✅ Direct (already solved) |
| `DiffBrightnessAdditive` | `offset` | `diff_brightness_additive` | `offset` | ✅ Direct (already solved) |

**Comment by xp1632** [comment]
### Solution

#### Diff Ops: Drop-in replacements that fix gradient flow

We created differentiable versions of each problematic operation. These are mathematically equivalent to the originals at initialization, but with learnable parameters that receive gradients.

| Diff Op (New) | What Changed | Learnable Params | Gradient Fix |
|---|---|---|---|
| `DiffGaussianBlur` | sigma is the learnable param; kernel_size auto-derived via 3σ rule | `sigma` [0.1, 10.0] | Gradients flow through Kornia's `exp(-x²/2σ²)` kernel construction |
| `DiffUnsharpMask` | Same sigma fix | `sigma` [0.1, 10.0] | Same as DiffGaussianBlur |
| `DiffSobel` | Wrapped as Operation (already differentiable in Kornia) | None (fixed kernel) | Gradients flow through image path |
| `DiffMorphology` | Manual `unfold` + SE addition + differentiable `max`/`min` | `se_values` [0.0, 1.0] (25 floats, 5×5) | SE values participate in computation graph via addition |
| `DiffThreshold` | Soft thresholding via sigmoid: `sigmoid(steepness × (input - threshold))` | `threshold` [0.01, 0.99], `steepness` [1.0, 50.0] | Sigmoid is smooth and differentiable everywhere |
| `DiffRgbToGrayscale` | Learnable channel weights via softmax normalization | `weights` [-5.0, 5.0] (3 floats, pre-softmax) | Softmax ensures weights sum to 1; gradients flow through |
| `DiffBrightness` | Multiplicative: `clamp(input × factor, 0, 1)` | `factor` [0.01, 2.0] | Direct multiplication, fully differentiable |
| `DiffContrast` | Mean-centered: `mean + (input - mean) × factor` | `factor` [0.01, 3.0] | Linear transform, fully differentiable |
| `DiffGamma` | Power function: `input^gamma` (with epsilon guard) | `gamma` [0.5, 2.5] | Power function is differentiable for positive inputs |
| `DiffSaturation` | Grayscale blend: `gray + (input - gray) × factor` | `factor` [0.0, 2.0] | Linear interpolation, fully differentiable |
| `DiffBrightnessAdditive` | Additive: `clamp(input + offset, 0, 1)` | `offset` [-0.5, 0.5] | Simple addition, fully differentiable |

#### Chaldene Visualization: Hybrid mapping strategy

For the animation that shows optimization progress on Chaldene visual nodes:

| Op | Chaldene Strategy | Reason |
|---|---|---|
| `DiffGaussianBlur` | → Map to Kornia `gaussian_blur2d` with sigma conversion: `ks = 2×⌈3σ⌉+1`, `sigma_x = sigma_y = sigma` | Clean mathematical conversion, same visual result |
| `DiffUnsharpMask` | → Map to Kornia `unsharp_mask` with same sigma conversion | Same as above |
| `DiffSobel` | → Map to Kornia `sobel` | No params to convert |
| `DiffMorphology` | → Map to Kornia `dilation`/`erosion`/etc. with fixed kernel size 5 | SE values can't be shown as a simple slider; keep static during animation |
| `DiffThreshold` | → New `diff_threshold` Chaldene NodeSpec needed | No Kornia equivalent exists |
| `DiffRgbToGrayscale` | → New `diff_rgb_to_grayscale` Chaldene NodeSpec needed | Kornia version has no learnable weights to show |
| `DiffBrightness/Contrast/Gamma/Saturation/BrightnessAdditive` | → Existing `diff_*` Chaldene NodeSpecs | Already implemented and working ✅ |

#### DiffMorphology: Paper visualization

**-Structuring Element:  binary mask ---> learnable weight map**

For the paper (not the Chaldene animation), the structuring element can be shown as a **heatmap**:

```
Initial SE (all 1s):          Optimized SE (learned):
┌─────────────────┐           ┌─────────────────┐
│ 1.0 1.0 1.0 1.0 1.0 │      │ 0.1 0.3 0.8 0.3 0.1 │
│ 1.0 1.0 1.0 1.0 1.0 │      │ 0.3 0.9 1.0 0.9 0.3 │
│ 1.0 1.0 1.0 1.0 1.0 │  →   │ 0.8 1.0 1.0 1.0 0.8 │
│ 1.0 1.0 1.0 1.0 1.0 │      │ 0.3 0.9 1.0 0.9 0.3 │
│ 1.0 1.0 1.0 1.0 1.0 │      │ 0.1 0.3 0.8 0.3 0.1 │
└─────────────────┘           └─────────────────┘
  (square, uniform)             (learned soft circle)
```

- Values near **1.0** = active morphological position (full effect)
- Values near **0.0** = inactive position (no effect at that offset)
- The SE naturally learns **task-appropriate shapes** (e.g., circular for isotropic morphology, elongated for directional)

#### Why original Kornia morphology cannot be differentiated

Two distinct reasons:

1. **No learnable parameters**: The structuring element is stored in `non_diff_params` as `torch.ones(5,5)` — the optimizer never sees it. `list(op.parameters()) == []`.

2. **Kornia's implementation doesn't propagate gradients through the SE**: Even if you made it a `nn.Parameter`, Kornia treats the SE kernel as a fixed mask for max-pool/min-pool. Our DiffMorphology uses manual `F.unfold` + explicit addition of SE values to the unfolded patches, ensuring the SE participates in the autograd computation graph.

**DiffMorphology formula**:
- Dilation: `out = max(input_patch + SE)` over neighborhood
- Erosion: `out = min(input_patch - SE)` over neighborhood

`max`/`min` are differentiable in PyTorch (gradients flow to the argmax/argmin element). The SE values get gradients via the `+`/`-` operations.

----

### Segmentation Pipeline (enabled by Diff ops)

With these differentiable operations, we can now optimize full segmentation pipelines end-to-end:

```
Input → DiffContrast → DiffGaussianBlur → DiffRgbToGrayscale → DiffThreshold → mask
                                                                                  |
                                                                         MSE vs target mask
```

Every operation in this chain is differentiable, so gradient descent can optimize all parameters simultaneously to produce the best segmentation mask.
